src/solc_0.8/OperatorFilterer/contracts/upgradeable/OperatorFiltererUpgradeable.sol

```diff
@@ -0,0 +1,56 @@
+//SPDX-License-Identifier: MIT
+pragma solidity 0.8.2;
+
+import {IOperatorFilterRegistry} from "../../interfaces/IOperatorFilterRegistry.sol";
+import {Initializable} from "@openzeppelin/contracts-upgradeable/proxy/utils/Initializable.sol";
+
+abstract contract OperatorFiltererUpgradeable is Initializable {
+    IOperatorFilterRegistry public operatorFilterRegistry =
+        IOperatorFilterRegistry(0x000000000000AAeB6D7670E522A718067333cd4E);
+
+    function __OperatorFilterer_init(address subscriptionOrRegistrantToCopy, bool subscribe) internal onlyInitializing {
+        // If an inheriting token contract is deployed to a network without the registry deployed, the modifier
+        // will not revert, but the contract will need to be registered with the registry once it is deployed in
+        // order for the modifier to filter addresses.
+        if (address(operatorFilterRegistry).code.length > 0) {
+            if (!operatorFilterRegistry.isRegistered(address(this))) {
+                if (subscribe) {
+                    operatorFilterRegistry.registerAndSubscribe(address(this), subscriptionOrRegistrantToCopy);
+                } else {
+                    if (subscriptionOrRegistrantToCopy != address(0)) {
+                        operatorFilterRegistry.registerAndCopyEntries(address(this), subscriptionOrRegistrantToCopy);
+                    } else {
+                        operatorFilterRegistry.register(address(this));
+                    }
+                }
+            }
+        }
+    }
+
+    modifier onlyAllowedOperator(address from) virtual {
+        // Check registry code length to facilitate testing in environments without a deployed registry.
+        if (address(operatorFilterRegistry).code.length > 0) {
+            // Allow spending tokens from addresses with balance
+            // Note that this still allows listings and marketplaces with escrow to transfer tokens if transferred
+            // from an EOA.
+            if (from == msg.sender) {
```

Comment:
ERC2771 ?

Reply:
same as above

Reply:
I thing we won't be allowed to exempt the Trusted forwarder as technically it is also an operator which needed to be checked by the OperatorFilterRegistry.